substrate/frame/identity/src/lib.rs

```diff
@@ -886,6 +966,181 @@ pub mod pallet {
+				&sender,
+				|maybe_authority| -> Result<Suffix, DispatchError> {
+					let properties =
+						maybe_authority.as_mut().ok_or(Error::<T>::NotUsernameAuthority)?;
+					ensure!(properties.allocation > 0, Error::<T>::NoAllocation);
+					properties.allocation.saturating_dec();
+					Ok(properties.suffix.clone())
+				},
+			)?;
+
+			// Verify input length before allocating a Vec with the user's input.
+			// `<` instead of `<=` because it needs one element for the point
+			// (`username` + `.` + `suffix`)
+			ensure!(
+				(username.len().saturating_add(suffix.len()) as u32) < USERNAME_MAX_LENGTH,
+				Error::<T>::InvalidUsername
+			);
+
+			// Ensure that the username only contains allowed characters. We already know the suffix
+			// does.
+			Self::validate_username(&username)?;
+
+			// Concatenate the username with suffix and cast into a BoundedVec. Should be infallible
+			// since we already ensured it is below the max length.
+			let mut full_username =
+				Vec::with_capacity(username.len().saturating_add(suffix.len()).saturating_add(1));
+			full_username.extend(username);
+			full_username.extend(b".");
+			full_username.extend(suffix);
+			let bounded_username =
+				Username::try_from(full_username).map_err(|_| Error::<T>::InvalidUsername)?;
+
+			// Usernames must be unique. Ensure it's not taken.
+			ensure!(
+				!AccountOfUsername::<T>::contains_key(bounded_username.clone()),
+				Error::<T>::UsernameTaken
+			);
+
+			// Insert or queue.
+			let who = T::Lookup::lookup(who)?;
+			if let Some(s) = signature {
+				// Account has pre-signed an authorization. Verify the signature provided and grant
+				// the username directly.
+				let encoded = Encode::encode(&bounded_username.to_vec());
+				Self::validate_signature(&encoded, &s, &who)?;
+				Self::insert_username(&who, bounded_username);
+			} else {
+				// The user must accept the username, therefore, queue it.
+				Self::queue_acceptance(&who, bounded_username);
+			}
+			Ok(())
+		}
+
+		/// Accept a given username that an `authority` granted. The call must include the full
+		/// username, as in `username.suffix`.
+		#[pallet::call_index(18)]
+		#[pallet::weight(T::WeightInfo::accept_username())]
+		pub fn accept_username(
+			origin: OriginFor<T>,
+			username: Username,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let (approved_for, _) =
+				PendingUsernames::<T>::take(&username).ok_or(Error::<T>::NoUsername)?;
+			ensure!(approved_for == who.clone(), Error::<T>::InvalidUsername);
+			Self::insert_username(&who, username.clone());
+			Self::deposit_event(Event::UsernameSet { who: who.clone(), username });
+			Ok(Pays::No.into())
+		}
+
+		/// Remove an expired username approval. The call must include the full username, as in
+		/// `username.suffix`.
+		#[pallet::call_index(19)]
+		#[pallet::weight(T::WeightInfo::remove_expired_approval())]
+		pub fn remove_expired_approval(
```

Comment:
I'm not a big fan of `on_idle` storage clearing, but since there's no deposit on the pre-approvals maybe we could have the clearing of expired pre approvals running on on_idle.

Reply:
It would need to select usernames though. I think we can write something offchain for this.

Reply:
Task API can be used for this.